semgrep-core/src/matching/Generic_vs_generic.ml

```diff
@@ -458,10 +461,15 @@ let rec m_name a b =
                };
              _;
            } as infob) ) ) -> (
+      let is_resolved =
+        match resolved with
+        | B.ResolvedName _ -> true
+        | _ -> false
+      in
       m_name a (B.Id (idb, { infob with B.id_resolved = ref None }))
       >||> try_alternate_names resolved
       (* Try the resolved entity *)
-      >||> m_name a (H.name_of_ids dotted)
+      >||> m_name ~is_resolved a (H.name_of_ids dotted)
```

Comment:
It could be more clear to just avoid `>||> m_name ~is_resolved a (H.name_of_ids dotted)` here in the first place. The way it is done works, but it's subtle, and I'm not sure it needs to be.  The way it is done for sym-prop is a lot more clear to me:

```ocaml
  | _, G.N (G.Id _) ->
      m_with_symbolic_propagation ~is_root (fun b1 -> m_expr a b1) b
```

Could not we replicate this here?

Reply:
Alternatively, we could have something like:

```ocaml
>||> (if is_metavariable a && is_resolved then
          (* Resolved names do not correspond to actual code, some tokens come from import declarations
           * whereas others come from uses of the imported entities, leading to non-sensical ranges.
           *  Metavariables should only be bound to actual code. *)
          fail ()
      else m_name a (H.name_of_ids dotted))
```

Reply:
I'm not entirely sure, because I'm afraid of a case which looks like `$A.x` or `x.$A`. In both cases, we will proceed to match, and in the latter we will succeed because of the match on the `name_last` (which is a good thing). In the former case, I'm afraid of the possibility of `$A` matching a prefix of multiple qualifiers (like z.y.x, matching `z.y`). 

I'd rather keep the logic as is and defer as late as possible, because _then_ we know for sure we _must_ be matching a metavariable to an `IdQualified` (not just a single id, which is fine) that is from a `ResolvedName`.